contrib/nextjs/defs.bzl

```diff
@@ -0,0 +1,317 @@
+    directory of the `nextjs()` macro.
+
+    Args:
+        name: the name of the application binary target
+
+        app: the nextjs application directory which was compiled in with `output: "standalone"
+
+            This is typically the output of the `nextjs()` macro.
+
+        deps: dependencies required to run the standalone application
+
+            This would normally include the `:node_modules/next` and `:node_modules/react` packages
+            that are not bundled into the standalone application.
+
+            This is separate from the `nextjs(deps)` required to bundle the application.
+
+        **kwargs: other attributes passed to the `js_binary` target such as `visibility`.
+    """
+
+    # The output directory containing the standalone application.
+    # TODO: make it configurable in the macro?
+    standalone_outdir = "%s-standalone" % name
+
+    # The subdirectory server.js is in within the standalone/ directory.
+    # TODO: how to override this in nextjs config? Or need to make it configurable in the macro?
+    pkg = native.package_name()
+
+    # The server binary and required environment
+    js_binary(
+        name = name,
+        entry_point = ":_{}.js".format(name),
+        chdir = native.package_name(),
+        data = deps,
+        **kwargs
+    )
+
+    # The server entry point into the standalone directory.
+    directory_path(
+        name = "_{}.js".format(name),
+        directory = ":_{}.standalone".format(name),
+        path = "standalone/{}/server.js".format(pkg),
+        visibility = ["//visibility:private"],
+        tags = ["manual"],
+    )
+
+    # Copy the standalone folder build and public/static to create a standalone server.
+    # See https://nextjs.org/docs/pages/api-reference/config/next-config-js/output#automatically-copying-traced-files
+    copy_to_directory(
+        name = "_{}.standalone".format(name),
+        srcs = [app] + native.glob(["public/**"]),
+        include_srcs_patterns = [
+            "public/**",
+            "{}/static/**".format(_next_build_out),
+            "{}/standalone/**".format(_next_build_out),
+        ],
+        exclude_srcs_patterns = [
+            # TODO: exclude non-deterministic and log/trace files?
+
+            # NOTE: the rules_js/contrib/nextjs config should have already excluded the node_modules
+        ],
+        replace_prefixes = {
+            "{}/standalone".format(_next_build_out): "standalone",
+            "{}/static".format(_next_build_out): "standalone/{}/{}/static".format(pkg, _next_build_out),
+            "public": "standalone/{}/public".format(pkg),
+        },
+        out = standalone_outdir,
+        visibility = ["//visibility:private"],
+        tags = ["manual"],
+    )
+
+# ---------------------------------------------------------------------------------------------
+# Next.js image layer - an optimal image layer for Next.js standalone applications.
```

Comment:
nit: needs to mention container images - many readers won't connect what a "layer" is without some context

Reply:
I'm dropping the image layer stuff from here for now. This is more generic "exclude packages from my image" then it is nextjs specific imo.